File: playground/BOJ/boj_4673_Self_number/boj_4673.js
/*----------------------------------------------
Sub  : [BOJ] 셀프 넘버
Link : https://www.acmicpc.net/problem/4673
Level: 실버 5
Tag  : JS, Brute Force
Memo
 - 
----------------------------------------------*/

const allNumberSet = new Set();
for (let i = 1; i < 10000; i++) {
  allNumberSet.add(i);
}

const createNumberSet = new Set();

allNumberSet.forEach(i => {
  let createNumber = i;
  for (const digit of String(i)) {
    createNumber += Number(digit);
  }
  createNumberSet.add(createNumber);
});

const selfNumberSet = [...allNumberSet].filter(num => !createNumberSet.has(num))
selfNumberSet.sort((a, b) => a - b);

// 결과 출력
selfNumberSet.forEach(num => console.log(num));

File: Online_Judge_Problems/boj/boj_4673_셀프넘버/boj_4673.c
/*----------------------------------------------
Sub  : [BOJ] 셀프 넘버
Link : https://www.acmicpc.net/problem/4673
Level: 실버 5
Tag  : C, Brute Force
Memo
 -
----------------------------------------------*/

#include <stdio.h>

#define MAX 10000

// 생성자를 이용해 숫자를 생성하는 함수
int generate(int n) {
    int sum = n;
    while (n > 0) {
        sum += n % 10;  // 각 자릿수를 더함
        n /= 10;
    }
    return sum;
}

int main() {
    int isSelfNumber[MAX];  // 1이면 셀프 넘버, 0이면 생성자가 있음

    // 배열을 모두 1(셀프 넘버)로 초기화
    for (int i = 0; i < MAX; i++) {
        isSelfNumber[i] = 1;
    }

    // 생성자가 있는 숫자를 체크
    for (int i = 1; i < MAX; i++) {
        int num = generate(i);
        if (num < MAX) {
            isSelfNumber[num] = 0;  // 생성자가 있는 숫자는 제외
        }
    }

    // 셀프 넘버 출력
    for (int i = 1; i < MAX; i++) {
        if (isSelfNumber[i] == 1) {
            printf("%d\n", i);
        }
    }

    return 0;
}
